app/src/main/java/com/andela/gkuti/mipista/service/LocationDetector.java

```diff
@@ -0,0 +1,179 @@
+     *
+     * @param context activity context
+     */
+    public LocationDetector(Context context) {
+        this.context = context;
+        intent = new Intent("location");
+    }
+
+    /**
+     * called when connection was successful
+     */
+    @Override
+    public void onConnected(@Nullable Bundle bundle) {
+        if (ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION)
+                == PackageManager.PERMISSION_GRANTED) {
+            LocationRequest locationRequest = LocationRequest.create();
+            locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
+            locationRequest.setInterval(60000);
+            LocationServices.FusedLocationApi.requestLocationUpdates(googleApiClient, locationRequest, this);
+        }
+    }
+
+    /**
+     * called when the connection was suspended
+     */
+    @Override
+    public void onConnectionSuspended(int i) {
+    }
+
+    /**
+     * called when the connection failed
+     */
+    @Override
+    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
+    }
+
+    /**
+     * method for setting up a connection
+     */
+    public void connect() {
+        country = " ";
+        state = " ";
+        locality = " ";
+        street = " ";
+        googleApiClient = new GoogleApiClient.Builder(context).addApi(LocationServices.API)
+                .addConnectionCallbacks(this)
+                .addOnConnectionFailedListener(this)
+                .build();
+        googleApiClient.connect();
+    }
+
+    /**
+     * method for disconnecting the connection
+     */
+    public void disconnect() {
+        try {
+            LocationServices.FusedLocationApi.removeLocationUpdates(
+                    googleApiClient, this);
+        } catch (Exception e) {
+        }
+
+    }
+
+    /**
+     * callback for changes in location
+     *
+     * @param location the new location
+     */
+    @Override
+    public void onLocationChanged(Location location) {
+        longitude = location.getLongitude();
+        latitude = location.getLatitude();
+        detectLocation();
+    }
+
+    /**
+     * method for detecting user location using Geocoder
+     */
+    private void detectLocation() {
+        Geocoder geocoder = new Geocoder(context);
+        List<Address> addresses;
+        try {
+            addresses = geocoder.getFromLocation(latitude, longitude, 1);
+            if (addresses != null) {
+                decodeAddress(addresses);
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        if (getAddress()) {
+            updateLocation();
+        }
+    }
+
+    /**
+     * method for getting the user location
+     *
+     * @return the return location else returns unknown
+     */
+    public String getLocation() {
+        if (getAddress()) {
+            return street + " " + locality + " " + state + " " + country;
```

Comment:
*Nitpick*: No comments on this . Just an interesting read 😄 --> http://stackoverflow.com/questions/47605/string-concatenation-concat-vs-operator